File: basic_dsa_four_hundred_two.java
/**
 * Basic DSA Problems - Remove K Digits (LeetCode 402)
 * Remove k digits from num to get smallest possible integer
 */
public class basic_dsa_four_hundred_two {

    // Problem: num string, k. Remove k digits. Return smallest result (as string).

    /**
     * Monotonic stack: keep smaller digits. Remove while top > current and k > 0.
     * Time O(n), Space O(n)
     */
    public static String removeKdigits(String num, int k) {
        if (k >= num.length()) return "0";
        java.util.Deque<Character> st = new java.util.ArrayDeque<>();
        for (char c : num.toCharArray()) {
            while (k > 0 && !st.isEmpty() && st.peek() > c) {
                st.pop();
                k--;
            }
            st.push(c);
        }
        while (k-- > 0) st.pop();
        StringBuilder sb = new StringBuilder();
        while (!st.isEmpty()) sb.append(st.pop());
        sb.reverse();
        while (sb.length() > 1 && sb.charAt(0) == '0') sb.deleteCharAt(0);
        return sb.length() == 0 ? "0" : sb.toString();
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Remove K Digits");
        System.out.println("=".repeat(60));

        System.out.println("1432219, k=3 -> " + removeKdigits("1432219", 3));
        System.out.println("10200, k=1 -> " + removeKdigits("10200", 1));
    }
}

File: basic_dsa_four_hundred_two.cpp
/**
 * Basic DSA - Validate Stack Sequences (LeetCode 946)
 * Simulate pushes in order; pop whenever top matches next wanted pop.
 */
#include <iostream>
#include <stack>
#include <string>
#include <vector>

bool validateStackSequences(const std::vector<int>& pushed, const std::vector<int>& popped) {
    std::stack<int> st;
    int j = 0;
    int m = static_cast<int>(popped.size());
    for (int x : pushed) {
        st.push(x);
        while (!st.empty() && j < m && st.top() == popped[j]) {
            st.pop();
            ++j;
        }
    }
    return j == m;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Validate Stack Sequences\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> p1 = {1, 2, 3, 4, 5};
    std::vector<int> o1 = {4, 5, 3, 2, 1};
    std::cout << "[1,2,3,4,5] push / [4,5,3,2,1] pop -> "
              << (validateStackSequences(p1, o1) ? "true" : "false") << "\n";

    std::vector<int> p2 = {1, 2, 3, 4, 5};
    std::vector<int> o2 = {4, 3, 5, 1, 2};
    std::cout << "[1,2,3,4,5] push / [4,3,5,1,2] pop -> "
              << (validateStackSequences(p2, o2) ? "true" : "false") << "\n";
}
